book/src/construction.md

```diff
@@ -35,13 +35,56 @@ An **Execution IR** (intermediate representation) is defined to describe the ste
 This IR is *not proved itself*; it only guides trace population.
 The actual soundness comes from the constraints inside the operation-specific STARK chips along with their lookups into the central witness table.
 
+In the IR, we differentiate between *primitive* and *nonprimitive operations*. 
+
+Primitive operations represent basic field arithmetic operations:
+
+- `Constant` -- for constants,
+- `Public` -- for public inputs,
+- `Add` -- for both addition and subtraction,
+- `Mul` -- for both multiplication and division. 
+
+They represent the core of the verification computation. Given only the primitive operations, one should be able to carry out most operations necessary in circuit verification. Primitive operations have the following properties:
+
+- They operate on elements of the witness table, through their `WitnessId` (index within the witness table).
+- The representation can be heavily optimized. For example, every time a constant is added to the IR, we wither create a new `WitnessId` or return an already existing one. We could also carry out common subexpression elimination.
+- They are executed in topological order during the circuit evaluation, and
+- they form a directed acyclic graph of dependencies.
+
+But relying only on primitive operations for the entire verification would lead to the introduction of many temporary values in the IR. In turn, this would lead to enlarged witness and primitive tables. This can be alleviated by the introduction of chips which carry out more complex operations: *nonprimitive operations*. These have more complex constraints, for example for sensitive cryptographic operations, and are verified by specialized AIRs. As such, they cannot be optimized similarly to primitive operations.
```

Comment:
actually you mention it here, so you can either tweak with my initial suggestion or ignore it :)